src/sinks/mqtt/sink.rs

```diff
@@ -0,0 +1,215 @@
+use std::{collections::VecDeque, fmt::Debug};
+
+use async_trait::async_trait;
+use bytes::BytesMut;
+use futures::{pin_mut, stream::BoxStream, Stream, StreamExt};
+use rumqttc::{AsyncClient, ClientError, ConnectionError, EventLoop, MqttOptions, QoS};
+use snafu::{ResultExt, Snafu};
+use tokio_util::codec::Encoder as _;
+use vector_common::internal_event::{ByteSize, CountByteSize, Output, Protocol, Registered};
+use vector_core::{
+    event::EventFinalizers,
+    internal_event::{BytesSent, EventsSent, InternalEventHandle},
+    ByteSizeOf,
+};
+
+use crate::sinks::mqtt::config::MqttQoS;
+use crate::{
+    codecs::{Encoder, Transformer},
+    emit,
+    event::{Event, EventStatus, Finalizable},
+    internal_events::TemplateRenderingError,
+    internal_events::{ConnectionOpen, MqttClientError, MqttConnectionError, OpenGauge},
+    sinks::mqtt::config::MqttSinkConfig,
+    sinks::util::StreamSink,
+    template::{Template, TemplateParseError},
+    tls::TlsError,
+};
+
+#[derive(Debug, Snafu)]
+#[snafu(visibility(pub))]
+pub enum MqttError {
+    #[snafu(display("invalid topic template: {}", source))]
+    TopicTemplate { source: TemplateParseError },
+    #[snafu(display("MQTT connection error: {}", source))]
+    Connection { source: ConnectionError },
+    #[snafu(display("TLS error: {}", source))]
+    Tls { source: TlsError },
+    #[snafu(display("MQTT client error: {}", source))]
+    Client { source: ClientError },
+}
+
+#[derive(Clone)]
+pub struct MqttConnector {
+    options: MqttOptions,
+    topic: Template,
+}
+
+impl MqttConnector {
+    pub fn new(options: MqttOptions, topic: String) -> Result<Self, MqttError> {
+        let topic = Template::try_from(topic).context(TopicTemplateSnafu)?;
+        Ok(Self { options, topic })
+    }
+
+    fn connect(&self) -> (AsyncClient, EventLoop) {
+        AsyncClient::new(self.options.clone(), 1024)
+    }
+
+    pub async fn healthcheck(&self) -> crate::Result<()> {
```

Comment:
This function doesn't seem to be doing anything useful. `self.connect` is not actually connecting it is just creating a client object. There's no error path to indicate if the healthcheck fails. 

Is there a way we can actually try connecting to the broker to see if it is listening?

Reply:
I do not see a way to perform the healthcheck in an easy way. I can imagine the scenario with a dedicated healthcheck topic, trying to publish into it a test event on a regular basis but... it seems to be an overengineering. @astro probably you have a suggestion.

@StephenWakely I think would be better just disable the healthcheck in MQTT sink since at least I see no way to implement it.

Reply:
FWIW there's a [Ping](https://docs.rs/rumqttc/latest/rumqttc/enum.Outgoing.html#variant.PingReq) in MQTT.

Reply:
Sure, I will check it. Just a bit later :)

Reply:
@StephenWakely unfortunately, I didn't find a way, how to send a Ping request. `client` is too high-level and allows just Publish usual packets and subscribe, nothing more.

With keeping alive probably will help - https://docs.rs/rumqttc/0.20.0/rumqttc/struct.MqttOptions.html#method.set_keep_alive

How to implement healthcheck - I do not know. Maybe would be easier just remove the healthcheck support from the sink.

Reply:
I've raised an issue on `rumqttc` to see how possible it will be https://github.com/bytebeamio/rumqtt/issues/562.

I think for now if you make the healthcheck always return `Ok(())` and then we will need to create an issue in Vector and update this healthcheck when we can.

Reply:
Sure. Will do it. I will create the corresponding issue after the merge of this PR.

Reply:
Fixed.